File: Heap/MinHeap.java
package org.surajmyt.Heap;

public class MinHeap {

    private int[] heap;
    private int size;
    private int max;

    public MinHeap(int max){
        this.size = 0;
        this.max = max;
        heap = new int[this.max];
    }

    public void insert(int node){
        if (size >= max){
            System.out.println("Insertion not possible");
            return;
        }

        heap[size] = node;

        int currentIdx = size;
        while (heap[currentIdx] < heap[parentNode(currentIdx)] ){
            swap(currentIdx, parentNode(currentIdx));
            currentIdx = parentNode(currentIdx);
        }
        size++;
    }

    public void minHeapfy(int currIdx){
        if (isLeafNode(currIdx)) return;

        int left = leftChild(currIdx);
        int right = rightChild(currIdx);

        while (heap[currIdx] > heap[left] || heap[currIdx] > heap[right]){
            if (heap[left] > heap[right]){
                swap(right, currIdx);
                minHeapfy(right);
            }
            else{
                swap(left, currIdx);
                minHeapfy(left);
            }
        }
    }

    public int removeElement(){
        if (size < 0) return -1;
        int remElement = heap[0];
        heap[0] = heap[--size];
        minHeapfy(0);
        return remElement;
    }

    private int parentNode(int index){return (index-1)/2;}

    private int leftChild(int index){return 2*index + 1;}

    private int rightChild(int index){return 2*index + 2;}

    private boolean isLeafNode(int index){
        return index > (size / 2) && index <= size;
    }

    public void swap(int index1, int index2){
        int temp = heap[index1];
        heap[index1] = heap[index2];
        heap[index2] = temp;
    }
    public void print(){
        for (int i = 0; i < size / 2; i++) {
            System.out.print("Parent : " + heap[i]);

            if (leftChild(i) < size)
                System.out.print(" Left Child :" + heap[leftChild(i)]);

            if (rightChild(i) < size)
                System.out.println(" Right Child: " + heap[rightChild(i)]);

            System.out.println();
        }
    }

    public static void main(String[] args) {
        MinHeap minHeap = new MinHeap(15);

        minHeap.insert(5);
        minHeap.insert(3);
        minHeap.insert(17);
        minHeap.insert(10);
        minHeap.insert(84);
        minHeap.insert(19);
        minHeap.insert(6);
        minHeap.insert(22);
        minHeap.insert(9);
        minHeap.print();

        System.out.println("Remove min root Node: " +minHeap.removeElement());
        minHeap.print();
    }
}

File: Heap/minHeap.ts
type MinHeap = {
  isEmpty: () => boolean;
  peek: () => node | undefined;
  push: (value: number) => boolean;
  pop: () => number | undefined;
};

type node = {
  value: number;
  left: node | undefined;
  right: node | undefined;
};

const createNode = (value: number): node => {
  return {
    value: value,
    left: undefined,
    right: undefined,
  };
};

const insertNode = (root: node, value: number): node => {
  const queue = [root];
  let index = 0;

  while (index < queue.length) {
    const poppedNode = queue[index]!;

    if (!poppedNode.left || !poppedNode.right) {
      if (!poppedNode.left) {
        poppedNode.left = createNode(value);
        queue.push(poppedNode.left);
      } else {
        poppedNode.right = createNode(value);
        queue.push(poppedNode.right);
      }

      return heapify(queue);
    }

    if (poppedNode.left) {
      queue.push(poppedNode.left);
    }
    if (poppedNode.right) {
      queue.push(poppedNode.right);
    }
    index++;
  }

  return root;
};

const heapify = (queue: node[]): node => {
  let newRoot = queue[0]!;
  let helperNode: node | undefined;

  const isChildMatch = (parent: node, value: number) => {
    return (
      (parent.left && value === parent.left!.value) ||
      (parent.right && value === parent.right!.value)
    );
  };

  while (queue.length > 0) {
    const poppedNode = queue.pop()!;

    if (queue.length > 0) {
      helperNode = queue.pop()!;

      while (!isChildMatch(helperNode, poppedNode.value)) {
        helperNode = queue.pop()!;
      }

      if (helperNode.value > poppedNode.value) {
        swapNodeValue(helperNode, poppedNode);

        if (poppedNode.value === newRoot.value) {
          return helperNode;
        }

        queue.push(helperNode);
      }
    } else {
      if (
        helperNode &&
        helperNode.value < poppedNode.value &&
        isChildMatch(poppedNode, helperNode.value)
      ) {
        swapNodeValue(helperNode, poppedNode);
      }

      return poppedNode;
    }
  }

  return newRoot;
};

const helperPop = (root: node): node | undefined => {
  const result = root.value;

  const queue = [root];
  let index = 0;
  let lastNode = root;

  while (index < queue.length) {
    const poppedNode = queue[index]!;

    if (poppedNode.left) {
      queue.push(poppedNode.left);
    }
    if (poppedNode.right) {
      queue.push(poppedNode.right);
    }

    index++;
  }

  lastNode = queue[queue.length - 1]!;

  if (lastNode.value === result) {
    return undefined;
  }

  root.value = lastNode.value;
  index = queue.length - 1;

  while (index >= 0) {
    const currentNode = queue[index]!;

    if (currentNode.left && currentNode.left.value === lastNode.value) {
      currentNode.left = undefined;
      break;
    }
    if (currentNode.right && currentNode.right.value === lastNode.value) {
      currentNode.right = undefined;
      break;
    }

    index--;
  }

  // check heap property violation for min heap
  let isHeapPropertyViolate = false;

  if (
    (root.left && root.left.value < root.value) ||
    (root.right && root.right.value < root.value)
  ) {
    isHeapPropertyViolate = true;
  }

  if (isHeapPropertyViolate) {
    heapifyDown(root);
  }

  return root;
};

function swapNodeValue(firstNode: node, secondNode: node) {
  const firstNodeValue = firstNode.value;
  firstNode.value = secondNode.value;
  secondNode.value = firstNodeValue;
}

function heapifyDown(currentNode: node) {
  while (currentNode !== null) {
    if (currentNode.left === undefined && currentNode.right === undefined) {
      break;
    } else if (
      currentNode.left === undefined &&
      currentNode.right !== undefined
    ) {
      if (currentNode.right!.value < currentNode.value) {
        swapNodeValue(currentNode, currentNode.right!);
        currentNode = currentNode.right;
      } else {
        break;
      }
    } else if (
      currentNode.right === undefined &&
      currentNode.left !== undefined
    ) {
      if (currentNode.left!.value < currentNode.value) {
        swapNodeValue(currentNode, currentNode.left!);
        currentNode = currentNode.left;
      } else {
        break;
      }
    } else {
      if (
        currentNode.left!.value < currentNode.right!.value &&
        currentNode.left!.value < currentNode.value
      ) {
        swapNodeValue(currentNode, currentNode.left!);
        currentNode = currentNode.left!;
      } else if (
        currentNode.left!.value > currentNode.right!.value &&
        currentNode.right!.value < currentNode.value
      ) {
        swapNodeValue(currentNode, currentNode.right!);
        currentNode = currentNode.right!;
      } else {
        break;
      }
    }
  }
}

export const minHeap = (): MinHeap => {
  let root: node | undefined = undefined;

  const isEmpty = (): boolean => {
    return root ? false : true;
  };

  const peek = (): node | undefined => {
    return root ? root : undefined;
  };

  const push = (value: number): boolean => {
    if (!root) {
      root = createNode(value);
      return true;
    }

    root = insertNode(root, value);
    return true;
  };

  const pop = (): number | undefined => {
    if (!root) {
      return undefined;
    }

    const poppedNode = root.value;
    root = helperPop(root);

    return poppedNode;
  };

  return {
    isEmpty,
    peek,
    push,
    pop,
  };
};

function main() {}

// main();
